### PR TITLE
perf: Replace `vmap`/one-hot `where_rectangular` and `select_per_label` with sort-based grouping using `offsets_from_counts`

### PR DESCRIPTION
**Pull request by n-gao**
Replace `jax.vmap` lookup and repeated key-sorting in `Index` with cache-aware validation and sort-based grouping

Key validation in `__post_init__` previously re-sorted the keys array on every construction. This is replaced with an `lru_cache`-backed `_check_index_keys` that validates once per unique `keys` tuple. Unhashable keys fall back to the unwrapped function so validation still runs without caching.

`select_per_label` and `where_rectangular` are rewritten to avoid patterns that scale poorly:

- `select_per_label` now handles the empty-index and single-label cases explicitly. For multiple labels it uses `offsets_from_counts` (imported alongside `subselect`) instead of recomputing cumulative offsets inline, and renames `sorted_indices` to `order` for clarity.
- `where_rectangular` replaces the `jax.vmap` over per-label `jnp.where(..., size=max_count)` calls — which allocated a separate scan per target key — with a single `jnp.argsort` pass followed by offset-indexed gather. An OOB bin is included in `bincount` so explicit sentinel rows are matched correctly. An early return handles the empty-index case.

Tests cover:
- Cached key validation still enforces uniqueness and sort order, and still checks index dtype, for a previously-seen `keys` tuple.
- Unhashable keys are validated without caching and reflect mutations.
- `select_per_label` with unequal populations, absent labels, out-of-bounds buffer entries, negative relative indices, and an empty index, across 1-, 4-, and 128-label cases.
- `where_rectangular` with repeated targets, missing keys, sentinel targets, and an empty source index.

### Performance measurements

16,384 shuffled index rows: rectangular lookup of 64 labels × 256 slots, and one-label selection with inactive rows. Key construction reuses a validated 16,384-key vocabulary.

| Workload | Trace, parent → PR (ms) | Runtime, parent → PR (ms) | Runtime speedup |
|---|---:|---:|---:|
| CPU rectangular lookup | 25.2 → 13.3 | 6.9 → 4.4 | 1.57× |
| L4 rectangular lookup | 25.5 → 13.5 | 0.216 → 0.231 | 0.93× |
| CPU single-label selection | 13.1 → 12.5 | 3.38 → 0.291 | 11.59× |
| L4 single-label selection | 12.8 → 12.8 | 0.237 → 0.122 | 1.94× |

Cached key construction on CPU: **5.41 → 0.00212 ms** per `Index` (15 batches of 100 constructions/process).

The rectangular lookup regresses on L4 at this size; the single-label path improves on both devices.

Parent `e183c9a` → PR `f493ad3`; JAX 0.10.0, Python 3.14, float64. CPU: Xeon 2.20 GHz, affinity 0–7; GPU: NVIDIA L4. Medians from two fresh processes/revision (parent→PR, then PR→parent): 3 cold traces and 25 warmed synchronous calls/process. Trace excludes lowering/XLA compilation; runtime includes dispatch. Speedup = parent/PR; below 1× is slower.

Larger-workload check: 131,072 rows; 128 labels × 1,024 slots for rectangular lookup.

| Workload | Runtime, parent → PR (ms) | Speedup | GPU utilization, parent → PR |
|---|---:|---:|---:|
| L4 rectangular lookup | 1.53 → 0.528 | 2.90× | 87% → 58% |
| L4 single-label selection | 0.392 → 0.127 | 3.09× | 61% → 18% |

Two fresh processes/revision, reversing order; each warmed phase lasts ≥3 s and ≥10 synchronous calls. Runtime uses flat array inputs/outputs to exclude host dataclass reconstruction. Utilization is `nvidia-smi` median at 200 ms intervals, omitting the first second; it measures device activity, not arithmetic occupancy. Numerical outputs agree across revisions.

### CODE REVIEW

**Comment by n-gao** [comment]
* **#313** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/313?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 2  dependent PRs ([#314](https://github.com/cusp-ai-oss/kUPS/pull/314) <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/314?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#315](https://github.com/cusp-ai-oss/kUPS/pull/315) <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/315?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#312** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/312?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#311** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/311?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#310** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/310?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/310?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

**Review by felixmusil — APPROVED** [review]

**Comment by n-gao** [comment]
### Merge activity

* **Sep 16, 9:28 AM UTC**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/310).
* **Sep 16, 9:28 AM UTC**: @n-gao merged this pull request with [Graphite](https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/310).
